torch/_inductor/coordinate_descent_tuner.py

```diff
@@ -80,27 +87,122 @@ def tunable_fields(self):
         return out
 
     @staticmethod
-    def get_neighbour_values(name, cur_val):
-        lhs_val = None
-        rhs_val = None
-        if name == "num_stages":
-            lhs_val = cur_val - 1
-            rhs_val = cur_val + 1
-        else:
-            lhs_val = cur_val // 2
-            rhs_val = cur_val * 2
+    def get_neighbour_values(name, orig_val, radius=1):
+        """
+        Get neighbour values in 'radius' steps. The original value is not
+        returned as it's own neighbour.
+        """
+        assert radius >= 1
+
+        def update(cur_val, inc=True):
+            if name == "num_stages":
+                if inc:
+                    return cur_val + 1
+                else:
+                    return cur_val - 1
+            else:
+                if inc:
+                    return cur_val * 2
+                else:
+                    return cur_val // 2
 
         out = []
-        if lhs_val > 0:
-            out.append(lhs_val)
-        out.append(rhs_val)
+        # increment loop
+        cur_val = orig_val
+        for _ in range(radius):
+            cur_val = update(cur_val, True)
+            out.append(cur_val)
+
+        # decrement loop
+        cur_val = orig_val
+        for _ in range(radius):
+            cur_val = update(cur_val, False)
+            if cur_val <= 0:
+                break
+            out.append(cur_val)
+
         return out
 
     @staticmethod
     def has_improvement(baseline, test):
         threshold = 0.001  # 0.1%
         return test is not None and test < baseline * (1 - threshold)
 
+    def check_all_dir(
+        self, func: Callable[[triton.Config], float], next_field_idx: int = 0
+    ):
+        """
+        Check all directions. We only do this once the regular coordinate
+        descent tuning find no better choices any more.
+        We only have a few tunable fields, so this should be fine.
+        """
+        if next_field_idx == len(self.tunable_fields):
+            # check if self.cur_config is better
+            # do deepcopy before running func since func may associate the
+            # config to the launcher.  If we do deepcopy after calling func,
+            # we may not able to find the launcher based on the copy of the config.
+            # That's because triton.Config uses class object's hash function.
+            config_copy = copy.deepcopy(self.cur_config)
+            cmp_res, candidate_timing = self.compare_config(
+                func, config_copy, self.best_config, self.best_timing
+            )
+            if cmp_res:
+                self.best_config = config_copy
+                self.best_timing = candidate_timing
+            else:
+                config_copy = None
+            return cmp_res
+
+        field = self.tunable_fields[next_field_idx]
+
+        # go thru different tuning choices for field
+        old_val = get_field(self.cur_config, field)
+
+        out = self.check_all_dir(func, next_field_idx + 1)
```

Comment:
This use of recursion is a bit hard to read.  It took me a second to realize it was iterating over `tunable_fields` in reverse order.  I'd suggest a simple loop.

Reply:
A simple loop may be hard to work here since the number of nested loops (i.e. len(tunable_fields) ) is not fixed.

Let me give an example, let's say the tunable_fields are XBLOCK, RBLOCK, num_warps (other kernels may not have RBLOCK, or they may have a YBLOCK), assuming the current best config we found is XBLOCK=64, RBLOCK=64, num_warps=16

We want to explore all neighbor settings. The neighbor values for XBLOCK = [32, 64, 128], YBLOCK=[32, 64, 128], num_warps=[8, 16, 32] (this assumes a radius = 1, a larger radius just checks more neighbor values for each field).
Then we will do a cross product of all these candidate neighbor values. 
-  for each (XBLOCK, YBLOCK, num_warps) in  [32, 64, 128] x [32, 64, 128] x [8, 16, 32]

we will generate a config and benchmark it.

Let me know if this part need more comments :)

Reply:
NVM, I think I can use `itertools.product` to replace the backtracking here.